### PR TITLE
xrRender: added stream buffers abstraction

### PR DESCRIPTION
**Pull request by vTurbine**

### CODE REVIEW

**Comment by vTurbine** [comment]
Todo (not in this pr):
- base classes for staging/stream buffers;
- stream buffer in `dx10ConstantBuffer.cpp`

File: src/Layers/xrRender/BufferUtils.h [hunk 1]
```diff
@@ -138,3 +138,95 @@ class IndexStagingBuffer
     u32 m_RefCounter{ 0 };
     bool m_AllowReadBack{ false }; // specifies whether host will want to have the data back (e.g. skinning code)
 };
+
+
+/**
+ * Stream buffer is another abstraction for mutable GPU data. It uses mapping to provide access
+ * to its content. The buffer data is considered to be write only.
+ *
+ * The are two modes of mapping available:
```
Comment (Xottab-DUTY) [inline src/Layers/xrRender/BufferUtils.h]:
```suggestion
 * There are two modes of mapping available:
```

File: src/Layers/xrRenderDX10/dx10BufferUtils.cpp [hunk 2]
```diff
@@ -16,51 +16,52 @@ u32 GetDeclLength(const VertexElement* decl)
...
+namespace BufferUtils
+{
+// TODO: replace by streaming buffer instance in `dx10ConstantBuffer`
+HRESULT CreateConstantBuffer(ConstantBufferHandle* ppBuffer, UINT DataSize)
+{
+    return ::CreateConstantBuffer(ppBuffer, DataSize);
 }
 };
 
+
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderDX10/dx10BufferUtils.cpp]:
```suggestion
```

File: src/Layers/xrRenderDX10/dx10BufferUtils.cpp [hunk 3]
```diff
@@ -16,51 +16,52 @@ u32 GetDeclLength(const VertexElement* decl)
     return D3DXGetDeclLength(decl);
 }
 
-namespace BufferUtils
-{
-HRESULT IC CreateBuffer(ID3DBuffer** ppBuffer, const void* pData, UINT DataSize, bool /*bImmutable*/, bool bIndexBuffer)
+static HRESULT CreateBuffer(ID3DBuffer** ppBuffer, const void* pData, UINT DataSize,
+    bool bImmutable, D3D_BIND_FLAG BufferType)
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderDX10/dx10BufferUtils.cpp]:
```suggestion
    bool bImmutable, D3D_BIND_FLAG bufferType)
```

File: src/Layers/xrRenderDX10/dx10BufferUtils.cpp [hunk 4]
```diff
@@ -16,51 +16,52 @@ u32 GetDeclLength(const VertexElement* decl)
...
 {
     D3D_BUFFER_DESC desc;
     desc.ByteWidth = DataSize;
-    // desc.Usage = bImmutable ? D3D_USAGE_IMMUTABLE : D3D_USAGE_DEFAULT;
-    desc.Usage = D3D_USAGE_DEFAULT;
-    desc.BindFlags = bIndexBuffer ? D3D_BIND_INDEX_BUFFER : D3D_BIND_VERTEX_BUFFER;
-    desc.CPUAccessFlags = 0;
+    desc.Usage = bImmutable ? D3D_USAGE_DEFAULT : D3D_USAGE_DYNAMIC;
+    desc.BindFlags = BufferType;
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderDX10/dx10BufferUtils.cpp]:
```suggestion
    desc.BindFlags = bufferType;
```

File: src/Layers/xrRenderDX10/dx10BufferUtils.cpp [hunk 5]
```diff
@@ -16,51 +16,52 @@ u32 GetDeclLength(const VertexElement* decl)
...
+static HRESULT CreateBuffer(ID3DBuffer** ppBuffer, const void* pData, UINT DataSize,
+    bool bImmutable, D3D_BIND_FLAG BufferType)
 {
     D3D_BUFFER_DESC desc;
     desc.ByteWidth = DataSize;
-    // desc.Usage = bImmutable ? D3D_USAGE_IMMUTABLE : D3D_USAGE_DEFAULT;
-    desc.Usage = D3D_USAGE_DEFAULT;
-    desc.BindFlags = bIndexBuffer ? D3D_BIND_INDEX_BUFFER : D3D_BIND_VERTEX_BUFFER;
-    desc.CPUAccessFlags = 0;
+    desc.Usage = bImmutable ? D3D_USAGE_DEFAULT : D3D_USAGE_DYNAMIC;
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderDX10/dx10BufferUtils.cpp]: Variable changes meaning. Let's don't do such things.

**Review by Xottab-DUTY — CHANGES_REQUESTED** [review]

File: src/Layers/xrRenderDX10/dx10BufferUtils.cpp (hunk 3 above)
Comment (vTurbine) [reply src/Layers/xrRenderDX10/dx10BufferUtils.cpp]: So why still `DataSize` then?

Comment (Xottab-DUTY) [reply src/Layers/xrRenderDX10/dx10BufferUtils.cpp]: We can rename it too then =)

File: src/Layers/xrRender/BufferUtils.h [hunk 6]
```diff
@@ -138,3 +138,95 @@ class IndexStagingBuffer
...
+    operator VertexBufferHandle() const
+    {
+        return m_DeviceBuffer;
+    }
+
+private:
+    void Destroy();
+
+    VertexBufferHandle m_DeviceBuffer;
+    u32 m_RefCounter{ 0 };
```
Comment (Xottab-DUTY) [inline src/Layers/xrRender/BufferUtils.h]:
```suggestion
    u32 m_RefCounter{};
```

File: src/Layers/xrRender/BufferUtils.h [hunk 7]
```diff
@@ -138,3 +138,95 @@ class IndexStagingBuffer
...
+    operator IndexBufferHandle() const
+    {
+        return m_DeviceBuffer;
+    }
+
+private:
+    void Destroy();
+
+    IndexBufferHandle m_DeviceBuffer;
+    u32 m_RefCounter{ 0 };
```
Comment (Xottab-DUTY) [inline src/Layers/xrRender/BufferUtils.h]:
```suggestion
    u32 m_RefCounter{};
```

File: src/Layers/xrRenderDX10/dx10BufferUtils.cpp [hunk 8]
```diff
@@ -16,48 +16,48 @@ u32 GetDeclLength(const VertexElement* decl)
...
 {
     D3D_BUFFER_DESC desc;
-    desc.ByteWidth = DataSize;
-    // desc.Usage = bImmutable ? D3D_USAGE_IMMUTABLE : D3D_USAGE_DEFAULT;
-    desc.Usage = D3D_USAGE_DEFAULT;
-    desc.BindFlags = bIndexBuffer ? D3D_BIND_INDEX_BUFFER : D3D_BIND_VERTEX_BUFFER;
-    desc.CPUAccessFlags = 0;
-    desc.MiscFlags = 0;
+    desc.ByteWidth      = dataSize;
+    desc.Usage          = bDynamic ? D3D_USAGE_DYNAMIC : D3D_USAGE_DEFAULT;
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderDX10/dx10BufferUtils.cpp:24]: I just realized that we can pass `D3D_USAGE` as an function argument, as we did it for `D3D_BIND_FLAG`.

File: src/Layers/xrRenderDX10/dx10BufferUtils.cpp [hunk 9]
```diff
@@ -354,3 +356,119 @@ size_t IndexStagingBuffer::GetVideoMemoryUsage() const
...
+
+    HW.stats_manager.decrement_stats_vb(m_DeviceBuffer);
+    _RELEASE(m_DeviceBuffer);
+}
+
+void* VertexStreamBuffer::Map(size_t offset, size_t size, bool flush /*= false*/)
+{
+    VERIFY(m_DeviceBuffer);
+
+    void *pData = nullptr;
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderDX10/dx10BufferUtils.cpp]:
```suggestion
    void* pData = nullptr;
```

File: src/Layers/xrRenderDX10/dx10BufferUtils.cpp [hunk 10]
```diff
@@ -354,3 +356,119 @@ size_t IndexStagingBuffer::GetVideoMemoryUsage() const
...
+void VertexStreamBuffer::Destroy()
+{
+    if (m_DeviceBuffer == nullptr)
+        return;
+
+    HW.stats_manager.decrement_stats_vb(m_DeviceBuffer);
+    _RELEASE(m_DeviceBuffer);
+}
+
+void* VertexStreamBuffer::Map(size_t offset, size_t size, bool flush /*= false*/)
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderDX10/dx10BufferUtils.cpp]:
```suggestion
void* VertexStreamBuffer::Map(size_t offset, size_t /*size/*, bool flush /*= false*/)
```

File: src/Layers/xrRender/BufferUtils.h [hunk 11]
```diff
@@ -138,3 +138,95 @@ class IndexStagingBuffer
     u32 m_RefCounter{ 0 };
     bool m_AllowReadBack{ false }; // specifies whether host will want to have the data back (e.g. skinning code)
 };
+
+
```
Comment (Xottab-DUTY) [inline src/Layers/xrRender/BufferUtils.h]:
```suggestion
```

File: src/Layers/xrRender/BufferUtils.h [hunk 12]
```diff
@@ -138,3 +138,95 @@ class IndexStagingBuffer
...
+ * Stream buffer is another abstraction for mutable GPU data. It uses mapping to provide access
+ * to its content. The buffer data is considered to be write only.
+ *
+ * There are two modes of mapping available:
+ * 1. Append (default)
+ *    All previous data kept untouched but can't be overwritten.
+ * 2. Flush
+ *    Previous buffer content is discarded.
+ */
+
```
Comment (Xottab-DUTY) [inline src/Layers/xrRender/BufferUtils.h]:
```suggestion
```

File: src/Layers/xrRender/BufferUtils.h (hunk 6 above)
Comment (vTurbine) [reply src/Layers/xrRender/BufferUtils.h]: See the same var in the class above

File: src/Layers/xrRenderDX10/dx10BufferUtils.cpp (hunk 8 above)
Comment (vTurbine) [reply src/Layers/xrRenderDX10/dx10BufferUtils.cpp:24]: Good~

File: src/Layers/xrRenderDX10/dx10BufferUtils.cpp [hunk 13]
```diff
@@ -354,3 +356,119 @@ size_t IndexStagingBuffer::GetVideoMemoryUsage() const
...
+void IndexStreamBuffer::Destroy()
+{
+    if (m_DeviceBuffer == nullptr)
+        return;
+
+    HW.stats_manager.decrement_stats_ib(m_DeviceBuffer);
+    _RELEASE(m_DeviceBuffer);
+}
+
+void* IndexStreamBuffer::Map(size_t offset, size_t size, bool flush /*= false*/)
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderDX10/dx10BufferUtils.cpp]:
```suggestion
void* IndexStreamBuffer::Map(size_t offset, size_t /*size*/, bool flush /*= false*/)
```

File: src/Layers/xrRenderDX10/dx10BufferUtils.cpp [hunk 14]
```diff
@@ -354,3 +356,119 @@ size_t IndexStagingBuffer::GetVideoMemoryUsage() const
...
+{
+    VERIFY(m_DeviceBuffer);
+
+    void *pData = nullptr;
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderDX10/dx10BufferUtils.cpp]:
```suggestion
    void* pData = nullptr;
```

File: src/Layers/xrRender/BufferUtils.h (hunk 6 above)
Comment (Xottab-DUTY) [reply src/Layers/xrRender/BufferUtils.h]: Should be removed too, but not in this PR =)

**Pull request merged**